### ISSUE TITLE
Improve efficiency of query samples (esp. for FDW's)

### ISSUE THREAD

**Issue by rafatower** [CartoDB/Windshaft-cartodb#1118]
When using carto-vl for rendering of points datasets, sampling is used by default.

Something like the following is sent in the map config from the client to the tiler:
```js
  "sample": {
    "num_rows": 1000,
    "include_columns": [
      "cartodb_id"
    ]
  }
```

See https://github.com/CartoDB/carto-vl/blob/81cc70f2a3d6d29082d8fc15b37dc8c1e4addcc0/src/client/windshaft.js#L362-L365

In turn, the tiler may run the following query to get a sample of the original wrapped query:
https://github.com/CartoDB/Windshaft-cartodb/blob/ebf373e6808aaaa30b34c122367838077316661d/lib/cartodb/utils/query-utils.js#L107-L115

This is particularly inefficient when running through FDW's. E.g:

In carto-vl client code:
```js
   // Define source
     const source = new carto.source.Dataset('cdb_fdw_example.yellow_tripdata');
```
(a single table, not even references to columns)

The query, as seen by the DB through an `EXPLAIN ANALYZE`:
```sql
EXPLAIN (analyze, verbose) WITH __cdb_rndseed AS (
             SELECT setseed(0.5)
         )
         SELECT "cartodb_id"
             FROM (
         SELECT
             "cartodb_id","the_geom_webmercator"
         FROM
             (SELECT "cartodb_id","the_geom_webmercator" FROM cdb_fdw_example.yellow_tripdata ) _cdb_epoch_transformation ) AS __cdb_query
             WHERE random() < 0.00006857135960823263
             LIMIT 1500;
```
results in:
```
                                                                      QUERY PLAN                                                                       
-------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=100.43..298.97 rows=1500 width=8) (actual time=317.505..61622.120 rows=1042 loops=1)
   Output: yellow_tripdata.cartodb_id
   ->  Foreign Scan on cdb_fdw_citus.yellow_tripdata  (cost=100.43..643502.07 rows=4861116 width=8) (actual time=317.503..61620.724 rows=1042 loops=1)
         Output: yellow_tripdata.cartodb_id
         Filter: (random() < '6.85713596082326e-05'::double precision)
         Rows Removed by Filter: 14582306
         Remote SQL: SELECT cartodb_id FROM carto_lite.yellow_tripdata
 Planning Time: 68.373 ms
 Execution Time: 61683.126 ms
(9 rows)
```

Obviously, this is not efficient at all, as it is first retrieving everything, then filtering most of the rows, which subverts the intention of the sampling.

 A possible solution may be to add an `OFFSET` clause (even an `OFFSET 0`, as it can act as "an optimization fence". See https://www.postgresql.org/docs/11/postgres-fdw.html#id-1.11.7.42.15 ).

**Comment by rafatower** [comment]
An important remark: a "sampling method" based on `random()` will never be pushed down to the foreign server, cause it is a `VOLATILE` function (just immutable functions can be pushed down).

**Comment by rafatower** [comment]
I also wonder why, given the map config, it is not considered a single table here:
https://github.com/CartoDB/Windshaft-cartodb/blob/ebf373e6808aaaa30b34c122367838077316661d/lib/cartodb/utils/query-utils.js#L102

and why it does not use the `TABLESAMPLE` method here:
https://github.com/CartoDB/Windshaft-cartodb/blob/ebf373e6808aaaa30b34c122367838077316661d/lib/cartodb/utils/query-utils.js#L123-L125

**Comment by rafatower** [comment]
I tested something like:

```
EXPLAIN (analyze, verbose) WITH __cdb_rndseed AS (
             SELECT setseed(0.5)
         ), __cdb_sampling_params AS (
           SELECT min(cartodb_id) AS min_id, max(cartodb_id) AS max_id, (max(cartodb_id) - min(cartodb_id)) AS id_span
           FROM
             (
               SELECT "cartodb_id","the_geom_webmercator" FROM cdb_fdw_example.yellow_tripdata
             ) _cdb_epoch_transformation
         ), __cdb_sampling_ids AS (
               SELECT p.min_id + trunc(random() * p.id_span)::integer AS cartodb_id
               FROM   __cdb_sampling_params p, generate_series(1, 1500) g
               GROUP  BY 1
         )
         SELECT * FROM (SELECT "cartodb_id","the_geom_webmercator" FROM cdb_fdw_example.yellow_tripdata) _cdb_epoch_transformation
         WHERE cartodb_id IN (SELECT cartodb_id FROM __cdb_sampling_ids)
         LIMIT 1500;
```
and although the sampling is done better, the join produces a chain of `PREPARE, DECLARE CURSOR, FETCH, CLOSE` and it ends up expending more total time in the query.

**Comment by rafatower** [comment]
The underlying assumption of that sampling method is that the query has a `cartodb_id` column.

After giving it much thought, I think the most cost-effective solution to the average and the big/huge table cases is to split it into two queries, using a similar approach to this [post](https://stackoverflow.com/questions/8674718/best-way-to-select-random-rows-postgresql):
- one for getting the min and max `cartodb_id` of the table/query
- another one with the pseudo-random ID's to filter, generated by javascript code.

This has the advantage of being robust and fast, as the max and min can be obtained very quickly thanks to the underlying index on `cartodb_id` and as there's no join between local and foreign, no `random` volatile function and no scan on function results.

**Comment by dgaubert** [comment]
Working on it!

If I recall correctly, the point to request a sample of data while instantiating the map is to have histograms, ramps, etc.. based on representative data. So the more randomized sample the better to get significant insights and create widgets outside the scope of CARTO VL.

Going to try the 2 steps proposal.

**Comment by dgaubert** [comment]
Implemented in #1120
